src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -112,142 +75,104 @@ public static bool TryMoveDown(Project project, IProjectTree projectTree)
-        /// </summary>
-        public static ImmutableArray<IProjectTree> GetAddedNodes(IProjectTree previousTarget, IProjectTree currentTarget)
-        {
-            Requires.NotNull(previousTarget, nameof(previousTarget));
-            Requires.NotNull(currentTarget, nameof(currentTarget));
+            var targetChild = GetChildren(target).FirstOrDefault();
+            if (targetChild == null)
+            {
+                targetChild = GetChildren(target.Parent).FirstOrDefault();
+                if (targetChild == null)
+                {
+                    return false;
+                }
+            }
 
-            return currentTarget.Children
-                .Where(updatedChild => IsAddedNode(previousTarget, updatedChild))
-                .OrderBy(GetDisplayOrder)
-                .ToImmutableArray();
+            var referenceElement = GetReferenceElement(project, targetChild, MoveAction.Above);
+            return TryMoveElements(elements, referenceElement, MoveAction.Above);
         }
 
         /// <summary>
-        /// This moves the nodes above the target's first child. 
-        /// If there are no children with a valid display order of the target or the target is an empty folder, it will try the parent target's first child.
+        /// Get project item elements based on the project tree.
+        /// Project tree can be a folder or item.
         /// </summary>
-        public static bool TryMoveNodesToTop(Project project, IEnumerable<IProjectTree> nodes, IProjectTree target)
+        public static ImmutableArray<ProjectItemElement> GetElements(Project project, IProjectTree projectTree)
         {
             Requires.NotNull(project, nameof(project));
-            Requires.NotNull(nodes, nameof(nodes));
-            Requires.NotNull(target, nameof(target));
-
-            // Move added nodes to the top. 
-            var nodeToAddAbove = GetFirstChild(target);
+            Requires.NotNull(projectTree, nameof(projectTree));
 
-            // This could have been an empty folder and nodeToAddAbove is null, therefore
-            //     get the first child of the target's parent.
-            // Empty folders do not have a valid display order.
-            // In the future, if we want to order empty folders, meaning they have a valid display order, 
-            //     this will need to be extended to handle that behavior.
+            var includes = GetEvaluatedIncludes(projectTree);
 
-            // This recursively tries to find a node to work with. If there is none at all, this will break on project root.
-            var currentTarget = target;
-            while(nodeToAddAbove == null && currentTarget.IsFolder && !GetChildren(currentTarget).Any())
-            {
-                currentTarget = currentTarget.Parent;
-                nodeToAddAbove = GetFirstChild(currentTarget);
-            }
+            var elements = new List<ProjectItemElement>();
```

Comment:
Use a ImmutableArray.Builder.

Reply:
Yes, great, that is what I definitely want. :)